src/number_theory/pythagorean_triples.lean

```diff
@@ -29,345 +223,330 @@ def circle_equiv_gen (hk : ∀ x : K, 1 + x^2 ≠ 0) :
+  { simpa [pythagorean_triple, neg_mul_neg] using h.eq, },
+  apply h'.is_primitive_classified_of_coprime_of_pos hc,
+  apply lt_of_le_of_ne _ (h'.ne_zero_of_coprime hc).symm,
+  exact le_neg.mp (not_lt.mp hz)
 end
 
-theorem exists_coprime_pythagorean_triple_pos {x y z : ℤ}
-  (h : pythagorean_triple x y z) (hc : int.gcd x y = 1) (hzpos : 0 < z):
-  (∃ (m n : ℤ), (x = m ^ 2 - n ^ 2 ∧ y = 2 * m * n))
-  ∨ ∃ (m n : ℤ), (x = 2 * m * n ∧ y = m ^ 2 - n ^ 2) :=
+theorem classified : h.is_classified :=
 begin
-  cases odd_even_of_coprime_pythagorean_triple h hc with h1 h2,
-  { apply or.intro_right, exact (exists_coprime_pythagorean_triple_of_odd h hc h1.right hzpos) },
-  apply or.intro_left, rw int.gcd_comm at hc,
-  have h2 : ∃ (m n : ℤ), y = 2 * m * n ∧ x = m ^ 2 - n ^ 2,
-  { exact (exists_coprime_pythagorean_triple_of_odd (pythagorean_triple_symm h) hc h2.left hzpos) },
-  tauto
+  by_cases h0 : int.gcd x y = 0,
+  { have hx : x = 0, { apply int.nat_abs_eq_zero.mp, apply nat.eq_zero_of_gcd_eq_zero_left h0 },
+    have hy : y = 0, { apply int.nat_abs_eq_zero.mp, apply nat.eq_zero_of_gcd_eq_zero_right h0 },
+    use [0, 1, 0], norm_num [hx, hy], },
+  apply h.is_classified_of_normalize_is_primitive_classified,
+  apply h.normalize.is_primitive_classified_of_coprime,
+  apply int.gcd_div_gcd_div_gcd (nat.pos_of_ne_zero h0),
 end
 
-theorem exists_coprime_pythagorean_triple {x y z : ℤ}
-  (h : pythagorean_triple x y z) (hc : int.gcd x y = 1) :
-  (∃ (m n : ℤ), (x = m ^ 2 - n ^ 2 ∧ y = 2 * m * n))
-  ∨ ∃ (m n : ℤ), (x = 2 * m * n ∧ y = m ^ 2 - n ^ 2) :=
+omit h
+
+theorem coprime_classification :
+  pythagorean_triple x y z ∧ int.gcd x y = 1 ↔
+  ∃ m n, ((x = m ^ 2 - n ^ 2 ∧ y = 2 * m * n) ∨
+            (x = 2 * m * n ∧ y = m ^ 2 - n ^ 2))
+          ∧ (z = m ^ 2 + n ^ 2 ∨ z = - (m ^ 2 + n ^ 2))
+          ∧ int.gcd m n = 1
+          ∧ ((m % 2 = 0 ∧ n % 2 = 1) ∨ (m % 2 = 1 ∧ n % 2 = 0)) :=
 begin
-  by_cases hz : 0 < z,
-  { exact exists_coprime_pythagorean_triple_pos h hc hz },
-  have h' : pythagorean_triple x y (-z),
-  { unfold pythagorean_triple, rw neg_mul_neg, exact h },
-  apply exists_coprime_pythagorean_triple_pos h' hc,
-  apply lt_of_le_of_ne _ (ne.symm (ne_zero_of_coprime_pythagorean_triple h' hc)),
-  exact le_neg.mp (not_lt.mp hz)
+  split,
+  { intro h,
+    obtain ⟨m, n, H⟩ := h.left.is_primitive_classified_of_coprime h.right,
+    use [m, n],
+    rcases H with ⟨⟨rfl, rfl⟩ | ⟨rfl, rfl⟩, co, pp⟩,
+    { refine ⟨or.inl ⟨rfl, rfl⟩, _, co, pp⟩,
+      have : z ^ 2 = (m ^ 2 + n ^ 2) ^ 2,
+      { rw [pow_two, ← h.left.eq], ring },
+      simpa using eq_or_eq_neg_of_pow_two_eq_pow_two _ _ this },
+    { refine ⟨or.inr ⟨rfl, rfl⟩, _, co, pp⟩,
+      have : z ^ 2 = (m ^ 2 + n ^ 2) ^ 2,
+      { rw [pow_two, ← h.left.eq], ring },
+      simpa using eq_or_eq_neg_of_pow_two_eq_pow_two _ _ this } },
+  { rintro ⟨m, n, ⟨rfl, rfl⟩ | ⟨rfl, rfl⟩, rfl | rfl, co, pp⟩;
+    delta pythagorean_triple,
+    { split, { ring },
```

Comment:
This seems like a good moment to use `all_goals` and/or `iterate` and/or `repeat`.